### PR TITLE
NIFI-1118 Update SplitText Processor - add support for split size lim…

### PR DESCRIPTION
**Pull request by jskora**
…its and header line markers.

1) Add "Maximum Fragment Size" property.  A new split file will be created if the next line to be added to the current split file exceeds this user-defined maximum file size.
2) Add "Header Line Marker Character" property.  Lines that begin with these user-defined character(s) will be considered header line(s) rather than a predetermined number of lines.  The existing property "Header Line Count" must be zero for this new property and behavior to be used.

### CODE REVIEW

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/SplitText.java [hunk 1]
```diff
@@ -75,6 +76,7 @@
 
     // attribute keys
     public static final String SPLIT_LINE_COUNT = "text.line.count";
+    public static final String SPLIT_SIZE = "fragment.size";
```
Comment (joewitt) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/SplitText.java]: recommend calling this "FRAGMENT_SIZE" to be consistent with the value and other fragment names

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/SplitText.java [hunk 2]
```diff
@@ -82,10 +84,16 @@
...
     public static final PropertyDescriptor LINE_SPLIT_COUNT = new PropertyDescriptor.Builder()
             .name("Line Split Count")
-            .description("The number of lines that will be added to each split file")
+            .description("The number of lines that will be added to each split file, excluding header lines")
             .required(true)
             .addValidator(StandardValidators.POSITIVE_INTEGER_VALIDATOR)
             .build();
+    public static final PropertyDescriptor FRAGMENT_MAX_SIZE = new PropertyDescriptor.Builder()
+            .name("Maximum Fragment Size")
+            .description("The maximum size of each split file, including header lines")
```
Comment (joewitt) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/SplitText.java]: Seems like there could be some interesting edge conditions.  For instance what if a given line is by itself larger than the max?  Or, does the max get exceeded by the size of at most one line then split or is the max exceeded then a pushback mechanism goes to the previous line and does the split and then continues?  I might be overthinking this but in any case would be good to let the user know how these edge cases are handled.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/SplitText.java (hunk 1 above)
Comment (jskora) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/SplitText.java]: Will do.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/SplitText.java (hunk 2 above)
Comment (jskora) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/SplitText.java]: I am checking that tests confirm the expected behavior, then I will clarify docs as needed.

**Comment by jskora** [comment]
I just pushed changes that address these issues, including tests for these and a couple more edge conditions.  It has been squashed (thanks @trkurc) into a single commit for readability.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/SplitText.java [hunk 3]
```diff
@@ -143,26 +165,12 @@ protected void init(final ProcessorInitializationContext context) {
...
-
-    private long countBytesToSplitPoint(final InputStream in, final OutputStream out, final boolean includeLineDelimiter) throws IOException {
+    private int readLine(final InputStream in, final OutputStream out,
+                          final boolean includeLineDelimiter) throws IOException {
         int lastByte = -1;
-        long bytesRead = 0L;
+        int bytesRead = 0;
 
         while (true) {
-            in.mark(1);
```
Comment (markap14) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/SplitText.java]: Trying to understand the logic here - why was this line removed? It looks like it is marked so that down below, on line 206 we can call in.reset() - with this removed, if we get to that line, where in.reset() is called, I believe it will thrown an IOException.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/SplitText.java [hunk 4]
```diff
@@ -198,23 +208,53 @@ private long countBytesToSplitPoint(final InputStream in, final OutputStream out
...
+                if (bytesRead == 0) {
+                    break;
+                }
+                info.lengthBytes += bytesRead;
+                info.lengthLines++;
+            }
+        // Else, keep reading all lines that begin with headerMarker character string
+        } else if (headerMarker != null) {
+            while (true) {
+                in.mark(0);
```
Comment (markap14) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/SplitText.java]: This call to in.mark(0) means that when in.reset() is called, we will reset back 0 bytes, essentially making the mark/reset do nothing. Not sure that I am following the logic here with how the mark/reset are being used.

Comment (markobean) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/SplitText.java]: This section of code is reading header lines a character at a time to determine if the line begins with the headerMarker character(s). After the line has been determined to be a header (or non-header), the buffer is reset so that a subsequent call to readLine() will capture all characters of the line. The maximum number of characters readHeader() will read before calling readLine() is the number of characters in the headerMarker String. Therefore, the in.mark(0) should be changed to in.mark(headerMarker.length).

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/SplitText.java [hunk 5]
```diff
@@ -230,100 +270,112 @@ public void onTrigger(final ProcessContext context, final ProcessSession session
...
+                                            flowFileInfo.lengthBytes = headerInfo.lengthBytes;
+                                            byteCount = headerInfo.lengthBytes;
                                             headerStream.writeTo(out);
-                                            linesCopied.set(readLines(in, splitCount, out, !removeTrailingNewlines));
+                                        }
+
+                                        // Process body
+                                        while (true) {
+                                            final ByteArrayOutputStream dataStream = new ByteArrayOutputStream();
+                                            in.mark(1);
```
Comment (markobean) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/SplitText.java]: This "in.mark()" is used to rollback the entire line if its inclusion in the current split exceeds maxFragmentSize. (Reset on line 336.) However, its readlimit should not be 1; it should be the maximum possible length of a line. Since the length of a given line is unknown, recommend using "Integer.MAX_VALUE."

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/SplitText.java (hunk 3 above)
Comment (markobean) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/SplitText.java]: This "in.mark(1)" was removed because marking and resetting the input stream has changed. Previously, this mark/reset was used to rollback the reading of the character after a \r. (More below.) Now, mark/reset is used at a higher level to potentially rollback an entire line read from the input flowfile if that line exceeds the size limit imposed by the FRAGMENT_MAX_SIZE property.
Removing this in.mark() does not generate an IOException in the below (line 206) in.reset() because an in.mark() was previously made prior to the call to readLine() (line 239 or line 323.) Nonetheless, the overall logic is still incorrect. There are two logical in.reset() conditions: character-based and line-based. This must be corrected and made consistent (line-based).
The special consideration of the \r character confuses me ('if' block beginning on Line 205.) In the original code, the byte after the \r is rolled back. Why? If the byte after \r is \n, the special consideration for \r is not required as the 'if' on line 194 captures the end of line. Is it valid and intended to have the \r indicate an end of line even when a subsequent \n is not present? (Note: testing of present SplitText processor shows incorrect behavior for only \r without \n; it duplicates the first character of the next "line".)
If special consideration for \r is not required and the 'if' block beginning on line 205 is removed, both the in.mark and in.reset within the readLine() method go away, and I believe all will be correct.

**Comment by jskora** [comment]
Cancelling this pull request, will followup with new PR for updated and current source branch.
